src/spec-lang/tc/typecheck.ts

```diff
@@ -779,6 +895,130 @@ function tcIdBuiltin(expr: SId): SType | undefined {
     return BuiltinSymbols.get(expr.name);
 }
 
+/**
+ * Given the type of some state variable `type`, a 'data-structure path' `path` find the path of the
+ * element of the data structre pointed to by the data-structure path.
+ *
+ * @param type
+ * @param path
+ */
+function locateElementType(type: TypeName, path: Array<SId | string>): SType {
+    for (let i = 0; i < path.length; i++) {
+        const element = path[i];
+
+        if (element instanceof SId) {
+            if (type instanceof ArrayTypeName) {
+                type = type.vBaseType;
+            } else if (type instanceof Mapping) {
+                type = type.vValueType;
+            } else {
+                throw new Error(
+                    `Mismatch between path ${pp(
+                        path
+                    )} and actual type at index ${i}: Expected indexable type but got ${pp(type)}`
+                );
+            }
+        } else {
+            if (
+                !(
+                    type instanceof UserDefinedTypeName &&
+                    type.vReferencedDeclaration instanceof StructDefinition
+                )
+            ) {
+                throw new Error(
+                    `Mismatch between path ${pp(
+                        path
+                    )} and actual type at index ${i}: Expected struct but got ${pp(type)}`
+                );
+            }
+
+            const structDef = type.vReferencedDeclaration;
+            const field = single(
+                structDef.vMembers.filter((def) => def.name === element),
+                `Expected a single field with name ${element} on struct  ${pp(structDef)}`
+            );
+
+            assert(
+                field.vType !== undefined,
+                `Missing type on field ${field.name} of struct ${pp(structDef)}`
+            );
+
+            type = field.vType;
+        }
+    }
+
+    return specializeType(astTypeNameToSType(type), DataLocation.Storage);
+}
+
+/**
+ * Given the type of some state variable `type`, a 'data-structure path' `path`, and an index `idx` in that path that
+ * corresponds to some index variable, find the type of that index variable.
+ *
+ * @param type
+ * @param idx
+ * @param path
```

Comment:
These line can be removed as do not provide anything useful.

Reply:
Fixed in dcf955c.